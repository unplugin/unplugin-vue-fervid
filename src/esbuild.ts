import { plugin } from "./core/index";

export default plugin.esbuild as typeof plugin.esbuild;
