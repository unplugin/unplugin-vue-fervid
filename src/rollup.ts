import { plugin } from './core'

export default plugin.rollup as typeof plugin.rollup
