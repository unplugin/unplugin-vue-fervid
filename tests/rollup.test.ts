import { resolve } from 'node:path'
import { describe, expect, it } from 'vitest'
import { rollup } from 'rollup'
import glob from 'fast-glob'
import ViteVue from '@vitejs/plugin-vue'
import * as vueCompiler from 'vue/compiler-sfc'
import esbuild from 'rollup-plugin-esbuild'
import Vue from '../src/rollup'
import type { Options } from '../src'

async function getCode(file: string, plugin: any) {
  const bundle = await rollup({
    input: [file],
    external: ['vue'],
    plugins: [plugin, esbuild({ format: 'esm' })],
  })
  const output = await bundle.generate({ format: 'esm' })
  return output.output
    .map((file) => {
      if (file.type === 'chunk') {
        return `//${file.fileName}\n${file.code}`
      } else {
        return file.fileName
      }
    })
    .join('\n')
}

function createPlugins(opt: Options) {
  const vite = ViteVue(opt)
  // @ts-expect-error
  vite.configResolved!({
    root: opt.root!,
    command: 'build',
    isProduction: opt.isProduction,
    build: {
      sourcemap: false,
    },
    define: {},
  } as any)
  return {
    unplugin: Vue(opt),
    vite,
  }
}

describe('transform', () => {
  describe('fixtures', async () => {
    const root = resolve(__dirname, '..')
    const files = await glob('tests/fixtures/*.{vue,js,ts}', {
      cwd: root,
      onlyFiles: true,
    })

    for (const file of files) {
      describe(file.replaceAll('\\', '/'), () => {
        const filepath = resolve(root, file)

        for (const isProduction of [true, false]) {
          it(`isProduction is ${isProduction}`, async () => {
            process.env.NODE_ENV = isProduction ? 'production' : 'development'

            const { unplugin, vite } = createPlugins({
              root,
              compiler: vueCompiler,
              isProduction,
            })

            const viteCode = await getCode(filepath, vite)
            const unpluginCode = await getCode(filepath, unplugin)

            expect(
              unpluginCode.replaceAll(JSON.stringify(filepath), "'#FILE#'"),
            ).toMatchSnapshot()
            expect(viteCode).toBe(unpluginCode)
          })
        }
      })
    }
  })
})
