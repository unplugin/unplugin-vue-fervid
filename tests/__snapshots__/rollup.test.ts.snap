// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transform > fixtures > tests/fixtures/basic.vue > isProduction is false 1`] = `
"//basic.js
import { openBlock, createElementBlock, createElementVNode } from 'vue';

var _export_sfc = (sfc, props) => {
  const target = sfc.__vccOpts || sfc;
  for (const [key, val] of props) {
    target[key] = val;
  }
  return target;
};

console.log('Hello world');

const _sfc_main = {};

const _hoisted_1 = /*#__PURE__*/createElementVNode("h1", null, "Hello World", -1 /* HOISTED */);
const _hoisted_2 = [
  _hoisted_1
];

function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return (openBlock(), createElementBlock("div", null, [..._hoisted_2]))
}
var basic = /*#__PURE__*/_export_sfc(_sfc_main, [['render',_sfc_render],['__file','#FILE#']]);

export { basic as default };
"
`;

exports[`transform > fixtures > tests/fixtures/basic.vue > isProduction is true 1`] = `
"//basic.js
import { openBlock, createElementBlock, createElementVNode } from 'vue';

var _export_sfc = (sfc, props) => {
  const target = sfc.__vccOpts || sfc;
  for (const [key, val] of props) {
    target[key] = val;
  }
  return target;
};

console.log('Hello world');

const _sfc_main = {};

const _hoisted_1 = /*#__PURE__*/createElementVNode("h1", null, "Hello World", -1 /* HOISTED */);
const _hoisted_2 = [
  _hoisted_1
];

function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return (openBlock(), createElementBlock("div", null, _hoisted_2))
}
var basic = /*#__PURE__*/_export_sfc(_sfc_main, [['render',_sfc_render]]);

export { basic as default };
"
`;

exports[`transform > fixtures > tests/fixtures/reactivity-transform.ts > isProduction is false 1`] = `
"//reactivity-transform.js
import { ref } from 'vue';

const ref1 = $ref("hello");
const ref2 = $(ref("hello"));
console.log(ref1);
console.log(ref2);
console.log($$(ref1));
console.log($$(ref2));
"
`;

exports[`transform > fixtures > tests/fixtures/reactivity-transform.ts > isProduction is true 1`] = `
"//reactivity-transform.js
import { ref } from 'vue';

const ref1 = $ref("hello");
const ref2 = $(ref("hello"));
console.log(ref1);
console.log(ref2);
console.log($$(ref1));
console.log($$(ref2));
"
`;

exports[`transform > fixtures > tests/fixtures/reactivity-transform.vue > isProduction is false 1`] = `
"//reactivity-transform.js
import { ref } from 'vue';

var _export_sfc = (sfc, props) => {
  const target = sfc.__vccOpts || sfc;
  for (const [key, val] of props) {
    target[key] = val;
  }
  return target;
};

const _sfc_main = {
  __name: 'reactivity-transform',
  setup(__props) {

const ref1 = $ref('hello');
const ref2 = $(ref('hello'));

console.log(ref1);
console.log(ref2);
console.log($$(ref1));
console.log($$(ref2));

return () => {}
}

};
var reactivityTransform = /*#__PURE__*/_export_sfc(_sfc_main, [['__file','#FILE#']]);

export { reactivityTransform as default };
"
`;

exports[`transform > fixtures > tests/fixtures/reactivity-transform.vue > isProduction is true 1`] = `
"//reactivity-transform.js
import { ref } from 'vue';

const _sfc_main = {
  __name: 'reactivity-transform',
  setup(__props) {

const ref1 = $ref('hello');
const ref2 = $(ref('hello'));

console.log(ref1);
console.log(ref2);
console.log($$(ref1));
console.log($$(ref2));

return () => {}
}

};

export { _sfc_main as default };
"
`;

exports[`transform > fixtures > tests/fixtures/script-setup.vue > isProduction is false 1`] = `
"//script-setup.js
import { openBlock, createElementBlock, createElementVNode } from 'vue';

var _export_sfc = (sfc, props) => {
  const target = sfc.__vccOpts || sfc;
  for (const [key, val] of props) {
    target[key] = val;
  }
  return target;
};

const _hoisted_1 = /*#__PURE__*/createElementVNode("h1", null, "Hello World", -1 /* HOISTED */);
const _hoisted_2 = [
  _hoisted_1
];

const _sfc_main = {
  __name: 'script-setup',
  setup(__props) {


return (_ctx, _cache) => {
  return (openBlock(), createElementBlock("div", null, _hoisted_2))
}
}

};
var scriptSetup = /*#__PURE__*/_export_sfc(_sfc_main, [['__file','#FILE#']]);

export { scriptSetup as default };
"
`;

exports[`transform > fixtures > tests/fixtures/script-setup.vue > isProduction is true 1`] = `
"//script-setup.js
import { openBlock, createElementBlock, createElementVNode } from 'vue';

const _hoisted_1 = /*#__PURE__*/createElementVNode("h1", null, "Hello World", -1 /* HOISTED */);
const _hoisted_2 = [
  _hoisted_1
];

const _sfc_main = {
  __name: 'script-setup',
  setup(__props) {


return (_ctx, _cache) => {
  return (openBlock(), createElementBlock("div", null, _hoisted_2))
}
}

};

export { _sfc_main as default };
"
`;

exports[`transform > fixtures > tests/fixtures/script-setup-ts.vue > isProduction is false 1`] = `
"//script-setup-ts.js
import { defineComponent, openBlock, createElementBlock, createElementVNode } from 'vue';

const _hoisted_1 = /* @__PURE__ */ createElementVNode(
  "h1",
  null,
  "Hello World",
  -1
  /* HOISTED */
);
const _hoisted_2 = [
  _hoisted_1
];
const str = "bar";
var script_setup_ts_vue_vue_type_script_setup_true_lang_default = /* @__PURE__ */ defineComponent({
  __name: "script-setup-ts",
  props: {
    foo: {},
    bar: {},
    baz: { type: Boolean }
  },
  setup(__props, { expose: __expose }) {
    __expose({
      str
    });
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock("div", null, _hoisted_2);
    };
  }
});

var _export_sfc = (sfc, props) => {
  const target = sfc.__vccOpts || sfc;
  for (const [key, val] of props) {
    target[key] = val;
  }
  return target;
};

var scriptSetupTs = /* @__PURE__ */ _export_sfc(script_setup_ts_vue_vue_type_script_setup_true_lang_default, [["__file", '#FILE#']]);

export { scriptSetupTs as default };
"
`;

exports[`transform > fixtures > tests/fixtures/script-setup-ts.vue > isProduction is true 1`] = `
"//script-setup-ts.js
import { defineComponent, openBlock, createElementBlock, createElementVNode } from 'vue';

const _hoisted_1 = /* @__PURE__ */ createElementVNode(
  "h1",
  null,
  "Hello World",
  -1
  /* HOISTED */
);
const _hoisted_2 = [
  _hoisted_1
];
const str = "bar";
var script_setup_ts_vue_vue_type_script_setup_true_lang_default = /* @__PURE__ */ defineComponent({
  __name: "script-setup-ts",
  props: {
    foo: {},
    bar: {},
    baz: { type: Boolean }
  },
  setup(__props, { expose: __expose }) {
    __expose({
      str
    });
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock("div", null, _hoisted_2);
    };
  }
});

export { script_setup_ts_vue_vue_type_script_setup_true_lang_default as default };
"
`;

exports[`transform > fixtures > tests/fixtures/sfc-src.vue > isProduction is false 1`] = `
"//sfc-src.js
import { defineComponent, openBlock, createElementBlock, createElementVNode, toDisplayString } from 'vue';

var src_script_ts_vue_type_script_src_true_lang_default = defineComponent({
  setup() {
    return [];
  }
});

const _hoisted_1 = /* @__PURE__ */ createElementVNode(
  "h1",
  null,
  "Hello World",
  -1
  /* HOISTED */
);
function render(_ctx, _cache, $props, $setup, $data, $options) {
  return openBlock(), createElementBlock("div", null, [
    _hoisted_1,
    createElementVNode(
      "h2",
      null,
      toDisplayString(_ctx.msg),
      1
      /* TEXT */
    )
  ]);
}

var _export_sfc = (sfc, props) => {
  const target = sfc.__vccOpts || sfc;
  for (const [key, val] of props) {
    target[key] = val;
  }
  return target;
};

var sfcSrc = /*#__PURE__*/_export_sfc(src_script_ts_vue_type_script_src_true_lang_default, [['render',render],['__file','#FILE#']]);

export { sfcSrc as default };
"
`;

exports[`transform > fixtures > tests/fixtures/sfc-src.vue > isProduction is true 1`] = `
"//sfc-src.js
import { defineComponent, openBlock, createElementBlock, createElementVNode, toDisplayString } from 'vue';

var src_script_ts_vue_type_script_src_true_lang_default = defineComponent({
  setup() {
    return [];
  }
});

const _hoisted_1 = /* @__PURE__ */ createElementVNode(
  "h1",
  null,
  "Hello World",
  -1
  /* HOISTED */
);
function render(_ctx, _cache, $props, $setup, $data, $options) {
  return openBlock(), createElementBlock("div", null, [
    _hoisted_1,
    createElementVNode(
      "h2",
      null,
      toDisplayString(_ctx.msg),
      1
      /* TEXT */
    )
  ]);
}

var _export_sfc = (sfc, props) => {
  const target = sfc.__vccOpts || sfc;
  for (const [key, val] of props) {
    target[key] = val;
  }
  return target;
};

var sfcSrc = /*#__PURE__*/_export_sfc(src_script_ts_vue_type_script_src_true_lang_default, [['render',render]]);

export { sfcSrc as default };
"
`;

exports[`transform > fixtures > tests/fixtures/src-script.ts > isProduction is false 1`] = `
"//src-script.js
import { defineComponent } from 'vue';

var src_script_default = defineComponent({
  setup() {
    return [];
  }
});

export { src_script_default as default };
"
`;

exports[`transform > fixtures > tests/fixtures/src-script.ts > isProduction is true 1`] = `
"//src-script.js
import { defineComponent } from 'vue';

var src_script_default = defineComponent({
  setup() {
    return [];
  }
});

export { src_script_default as default };
"
`;

exports[`transform > fixtures > tests/fixtures/typescript.vue > isProduction is false 1`] = `
"//typescript.js
var typescript_vue_vue_type_script_lang_default = {};

var _export_sfc = (sfc, props) => {
  const target = sfc.__vccOpts || sfc;
  for (const [key, val] of props) {
    target[key] = val;
  }
  return target;
};

var typescript = /* @__PURE__ */ _export_sfc(typescript_vue_vue_type_script_lang_default, [["__file", '#FILE#']]);

export { typescript as default };
"
`;

exports[`transform > fixtures > tests/fixtures/typescript.vue > isProduction is true 1`] = `
"//typescript.js
var typescript_vue_vue_type_script_lang_default = {};

export { typescript_vue_vue_type_script_lang_default as default };
"
`;
